src/msw/dirdlg.cpp

```diff
@@ -378,77 +383,122 @@ int wxDirDialog::ShowIFileDialog(WXHWND owner)
+                                 wxArrayString& paths)
+{
+    HRESULT hr = S_OK;
     wxString path;
+    wxArrayString tempPaths;
 
-    hr = fileDialog->Show(owner);
-    if ( SUCCEEDED(hr) )
+    if ( multipleSelection )
     {
-        wxCOMPtr<IShellItem> folder;
+        wxCOMPtr<IShellItemArray> itemArray;
 
-        hr = fileDialog->GetResult(&folder);
-        if ( SUCCEEDED(hr) )
+        hr = fileDialog->GetResults(&itemArray);
+        if ( FAILED(hr) )
         {
-            LPOLESTR pathOLE = NULL;
+            wxLogApiError(wxS("IShellItemArray::GetResults"), hr);
+            return false;
+        }
+
+        DWORD count = 0;
 
-            hr = folder->GetDisplayName(SIGDN_FILESYSPATH, &pathOLE);
-            if ( SUCCEEDED(hr) )
+        hr = itemArray->GetCount(&count);
+        if ( FAILED(hr) )
+        {
+            wxLogApiError(wxS("IShellItemArray::GetCount"), hr);
+            return false;
+        }
+
+        for ( DWORD i = 0; i < count; ++i )
+        {
+            wxCOMPtr<IShellItem> item;
+
+            hr = itemArray->GetItemAt(i, &item);
+            if ( FAILED(hr) )
             {
-                path = pathOLE;
-                CoTaskMemFree(pathOLE);
+                // do not attempt to retrieve any other items
+                // and just fail
+                wxLogApiError(wxS("IShellItemArray::GetItem"), hr);
+                tempPaths.clear();
+                break;
             }
-            else
+
+            if ( !ConvertIShellItemToPath(item, path) )
             {
-                wxLogApiError(wxS("IShellItem::GetDisplayName"), hr);
+                // again, just fail
+                tempPaths.clear();
+                break;
             }
+
+            tempPaths.push_back(path);
         }
-        else
-        {
-            wxLogApiError(wxS("IFileDialog::GetResult"), hr);
-        }
-    }
-    else if ( hr == HRESULT_FROM_WIN32(ERROR_CANCELLED) )
-    {
-        return wxID_CANCEL; // the user cancelled the dialog
+
     }
-    else
+    else // single selection
     {
-        wxLogApiError(wxS("IFileDialog::Show"), hr);
-    }
+        wxCOMPtr<IShellItem> item;
 
-    if ( path.empty() )
-    {
-        // the user didn't cancel the dialog and yet the path is empty
-        // it means there was an error, already logged by wxLogApiError()
-        // now report the error to the user and return
-        wxLogSysError(_("Couldn't obtain folder name"), hr);
-        return wxID_CANCEL;
+        hr = fileDialog->GetResult(&item);
+        if ( FAILED(hr) )
+        {
+            wxLogApiError(wxS("IFileOpenDialog::GetResult"), hr);
+            return false;
+        }
+
+        if ( !ConvertIShellItemToPath(item, path) )
+        {
+            return false;
+        }
+
+        tempPaths.push_back(path);
     }
 
-    m_path = path;
-    return wxID_OK;
+    if ( tempPaths.empty() )
+        return false; // there was en error
+
+    paths = tempPaths;
```

Comment:
Could do
```suggestion
    paths.swap(tempPaths);
```
for a tiny efficiency gain. Not really sure if it's worth it.

Reply:
I am aware of the trick, but I preferred to put emphasis on intent, i.e., one direction assignment and not a swap, believing the possible performance benefit negligible.